core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -98,22 +112,57 @@ public void processExpiration() {
                      expiredMortal = ExpiryHelper.isExpiredMortal(lifespan, ice.getCreated(), currentTimeMillis);
                      expiredTransient = ExpiryHelper.isExpiredTransient(maxIdle, ice.getLastUsed(), currentTimeMillis);
                   }
-                  // We check lifespan first as this is much less expensive to remove than max idle.
-                  if (expiredMortal) {
-                     addAndWaitIfFull(handleLifespanExpireEntry(ice.getKey(), value, lifespan, false), futures);
-                  } else if (expiredTransient) {
-                     addAndWaitIfFull(actualRemoveMaxIdleExpireEntry(ice.getKey(), value, maxIdle, false), futures);
+                  if (expiredMortal || expiredTransient) {
+                     if (expirationPermits.tryAcquire(1, TimeUnit.MINUTES)) {
+                        removedEntries++;
+                        CompletableFuture<?> stage;
+                        // We check lifespan first as this is much less expensive to remove than max idle.
```

Comment:
```suggestion
                        // If the entry is expired both wrt lifespan and wrt maxIdle, we perform lifespan expiration as it is cheaper
```
Question: could `handleLifespanExpireEntry` keep the entry in the cache, and if it does, should we also try the max-idle expiration?

Reply:
No, `handleLifespanExpireEntry`  will always remove it, unless the object has been written to concurrently. In that case you don't want to do any expiration anyways.